<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function, for each sample, will independently compute sums of score for each tile
per cell and then infer gene topic scores.
This function is adapted from addGeneScoreMatrix function from the  ArchR package."><title>Add TopicScoreMatrix to ArrowFiles or an ArchRProject — scoreGeneByPeak • Phoenix</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Add TopicScoreMatrix to ArrowFiles or an ArchRProject — scoreGeneByPeak"><meta property="og:description" content="This function, for each sample, will independently compute sums of score for each tile
per cell and then infer gene topic scores.
This function is adapted from addGeneScoreMatrix function from the  ArchR package."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Phoenix</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Add TopicScoreMatrix to ArrowFiles or an ArchRProject</h1>
      
      <div class="d-none name"><code>scoreGeneByPeak.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function, for each sample, will independently compute sums of score for each tile
per cell and then infer gene topic scores.
This function is adapted from <code>addGeneScoreMatrix</code> function from the  <code>ArchR</code> package.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scoreGeneByPeak</span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scoreMat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  geneModel <span class="op">=</span> <span class="st">"exp(-abs(x)/5000) + exp(-1)"</span>,</span>
<span>  peakWidth <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"sum"</span>,</span>
<span>  extendUpstream <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">1e+05</span><span class="op">)</span>,</span>
<span>  extendDownstream <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">1e+05</span><span class="op">)</span>,</span>
<span>  geneUpstream <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  geneDownstream <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  useGeneBoundaries <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useTSS <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  extendTSS <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  geneScaleFactor <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  excludeChr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrY"</span>, <span class="st">"chrM"</span><span class="op">)</span>,</span>
<span>  blacklist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subThreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  tstart <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  logFile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>genes</dt>
<dd><p>A stranded <code>GRanges</code> object containing the ranges associated with all gene start and end coordinates.</p></dd>


<dt>peaks</dt>
<dd><p>A <code>GRanges</code> object containing the peaks with a <code>score</code> column to calculate gene-level scores.</p></dd>


<dt>scoreMat</dt>
<dd></dd>


<dt>geneModel</dt>
<dd><p>A string giving a "gene model function" used for weighting peaks for gene score calculation. This string
should be a function of <code>x</code>, where <code>x</code> is the stranded distance from the transcription start site of the gene.</p></dd>


<dt>extendUpstream</dt>
<dd><p>The minimum and maximum number of basepairs upstream of the transcription start site to consider for gene
activity score calculation.</p></dd>


<dt>extendDownstream</dt>
<dd><p>The minimum and maximum number of basepairs downstream of the transcription start site or transcription termination site
(based on 'useTSS') to consider for gene activity score calculation.</p></dd>


<dt>geneUpstream</dt>
<dd><p>An integer describing the number of bp upstream the gene to extend the gene body. This effectively makes the gene body larger as there
are proximal peaks that should be weighted equally to the gene body. This parameter is used if 'useTSS=FALSE'.</p></dd>


<dt>geneDownstream</dt>
<dd><p>An integer describing the number of bp downstream the gene to extend the gene body.This effectively makes the gene body larger as there
are proximal peaks that should be weighted equally to the gene body. This parameter is used if 'useTSS=FALSE'.</p></dd>


<dt>useGeneBoundaries</dt>
<dd><p>A boolean value indicating whether gene boundaries should be employed during gene activity score
calculation. Gene boundaries refers to the process of preventing tiles from contributing to the gene score of a given gene
if there is a second gene's transcription start site between the tile and the gene of interest.</p></dd>


<dt>useTSS</dt>
<dd><p>A boolean describing whether to build gene model based on gene TSS or the gene body.</p></dd>


<dt>extendTSS</dt>
<dd><p>A boolean describing whether to extend the gene TSS. By default useTSS uses the 1bp TSS while this parameter enables the extension of this
region with 'geneUpstream' and 'geneDownstream' respectively.</p></dd>


<dt>geneScaleFactor</dt>
<dd><p>A numeric scaling factor to weight genes based on the inverse of there length i.e. (Scale Factor)/(Gene Length). This
is scaled from 1 to the scale factor. Small genes will be the scale factor while extremely large genes will be closer to 1. This scaling helps with
the relative gene score value.</p></dd>


<dt>excludeChr</dt>
<dd><p>A character vector containing the <code>seqnames</code> of the chromosomes that should be excluded from this analysis.</p></dd>


<dt>blacklist</dt>
<dd><p>A <code>GRanges</code> object containing genomic regions to blacklist that may be extremeley over-represented and thus
biasing the geneScores for genes nearby that locus.</p></dd>


<dt>subThreads</dt>
<dd></dd>


<dt>tstart</dt>
<dd></dd>


<dt>logFile</dt>
<dd><p>The path to a file to be used for logging ArchR output.</p></dd>


<dt>matrixName</dt>
<dd><p>The name to be used for storage of the gene activity score matrix in the provided <code>ArchRProject</code> or ArrowFiles.</p></dd>


<dt>force</dt>
<dd><p>A boolean value indicating whether to force the matrix indicated by <code>matrixName</code> to be overwritten if it already exist in the given <code>input</code>.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zepeng Mu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

